src/internal/pfsdb/commit_provenance.go

```diff
@@ -0,0 +1,225 @@
+           UNION ALL
+            SELECT cp.from_id, cp.to_id
+            FROM prov p, pfs.commit_provenance cp
+            WHERE cp.from_id = p.to_id
+          )
+          SELECT DISTINCT commit_id 
+          FROM pfs.commits, prov 
+          WHERE int_id = prov.to_id AND commit_set_id != $2;`
+	rows, err := tx.QueryxContext(ctx, q, id, id)
+	if err != nil {
+		return nil, errors.EnsureStack(err)
+	}
+	defer rows.Close()
+	cs := make([]*pfs.Commit, 0)
+	for rows.Next() {
+		var commit string
+		if err := rows.Scan(&commit); err != nil {
+			return nil, errors.EnsureStack(err)
+		}
+		cs = append(cs, ParseCommit(commit))
+	}
+	return cs, nil
+}
+
+// CommitSetSubvenance returns all the commit IDs that contain commits in this commit set in their
+// full provenance
+func CommitSetSubvenance(ctx context.Context, tx *pachsql.Tx, id string) ([]*pfs.Commit, error) {
+	q := `
+          WITH RECURSIVE subv(from_id, to_id) AS (
+            SELECT from_id, to_id 
+            FROM pfs.commit_provenance JOIN pfs.commits ON int_id = to_id 
+            WHERE commit_set_id = $1
+           UNION ALL
+            SELECT cp.from_id, cp.to_id
+            FROM subv s, pfs.commit_provenance cp
+            WHERE cp.to_id = s.from_id
+          )
+          SELECT DISTINCT commit_id 
+          FROM pfs.commits, subv 
+          WHERE int_id = subv.from_id AND commit_set_id != $2;`
+	rows, err := tx.QueryxContext(ctx, q, id, id)
+	if err != nil {
+		return nil, errors.EnsureStack(err)
+	}
+	defer rows.Close()
+	cs := make([]*pfs.Commit, 0)
+	for rows.Next() {
+		var commit string
+		if err := rows.Scan(&commit); err != nil {
+			return nil, errors.EnsureStack(err)
+		}
+		cs = append(cs, ParseCommit(commit))
+	}
+	return cs, nil
+}
+
+func AddCommit(ctx context.Context, tx *pachsql.Tx, commit *pfs.Commit) error {
+	stmt := `INSERT INTO pfs.commits(commit_id, commit_set_id) VALUES ($1, $2);`
+	_, err := tx.ExecContext(ctx, stmt, CommitKey(commit), commit.ID)
+	return errors.Wrapf(err, "insert commit %q into pfs.commits", CommitKey(commit))
+}
+
+func DeleteCommit(ctx context.Context, tx *pachsql.Tx, commitKey string) error {
+	id, err := getCommitTableID(ctx, tx, commitKey)
+	if err != nil {
+		return err
+	}
+	stmt := `DELETE FROM pfs.commits WHERE int_id = $1;`
+	_, err = tx.ExecContext(ctx, stmt, id)
+	return errors.Wrapf(err, "delete from pfs.commits")
+}
+
+func getCommitTableID(ctx context.Context, tx *pachsql.Tx, commitKey string) (int, error) {
+	query := `SELECT int_id FROM pfs.commits WHERE commit_id = $1;`
+	rows, err := tx.QueryxContext(ctx, query, commitKey)
+	if err != nil {
+		return 0, errors.EnsureStack(err)
+	}
+	var id int
+	for rows.Next() {
+		if err := rows.Scan(&id); err != nil {
+			return 0, errors.EnsureStack(err)
+		}
+	}
+	return id, nil
+}
+
+func AddCommitProvenance(ctx context.Context, tx *pachsql.Tx, from, to *pfs.Commit) error {
+	query := `SELECT int_id, commit_id FROM pfs.commits WHERE commit_id = $1 OR commit_id = $2;`
```

Comment:
It seems like the below logic would be much clearer if we just split this into two queries.

Reply:
for sure. this falls under premature optimization